src/main/java/com/mariamkatamashvlii/gym/service/implementation/TrainerServiceImpl.java

```diff
@@ -0,0 +1,194 @@
+
+        User user = userRepo.findByUsername(username);
+        if (user == null) {
+            log.error("[{}] User not found with username: {}", transactionId, username);
+            throw new EntityNotFoundException("User not found for username: " + username);
+        }
+
+        Trainer trainer = trainerRepo.findByUsername(username);
+        if (trainer == null) {
+            log.error("[{}] Trainer not found with username: {}", transactionId, username);
+            throw new EntityNotFoundException("Trainer not found for username: " + username);
+        }
+
+        user.setFirstName(updateRequestDTO.getFirstName());
+        user.setLastName(updateRequestDTO.getLastName());
+        user.setIsActive(updateRequestDTO.getIsActive());
+        userRepo.save(user);
+        TrainingTypeDTO specialization = new TrainingTypeDTO(
+                trainer.getSpecialization().getId(),
+                trainer.getSpecialization().getTrainingTypeName()
+        );
+        List<TraineeDTO> trainees = trainer.getTrainees().stream().map(trainee -> new TraineeDTO(
+                trainee.getUser().getUsername(),
+                trainee.getUser().getFirstName(),
+                trainee.getUser().getLastName()
+        )).toList();
+
+        log.info("[{}] Profile updated successfully for username: {}", transactionId, username);
+        return new UpdateResponseDTO(
+                user.getUsername(),
+                user.getFirstName(),
+                user.getLastName(),
+                specialization,
+                user.getIsActive(),
+                trainees
+        );
+    }
+
+    @Override
+    public List<TrainingResponseDTO> getTrainings(TrainerTrainingsRequestDTO trainerTrainingsRequestDTO) {
+        String transactionId = UUID.randomUUID().toString();
+        String username = trainerTrainingsRequestDTO.getUsername();
+        LocalDate fromDate = trainerTrainingsRequestDTO.getFromDate();
+        LocalDate toDate = trainerTrainingsRequestDTO.getToDate();
+        String traineeName = trainerTrainingsRequestDTO.getTraineeName();
+        log.info("[{}] Fetching trainings for trainer: {}", transactionId, username);
+
+        Trainer trainer = trainerRepo.findByUsername(username);
+        if (trainer == null || trainer.getTrainings() == null) {
+            log.info("No trainings found or trainer does not exist for username: {}", username);
+            return List.of();
+        }
+        log.info("[{}] Successfully fetched trainings for trainer: {}", transactionId, username);
+        return trainer.getTrainings().stream()
+                .filter(t -> (fromDate == null || !t.getTrainingDate().isBefore(fromDate)) && (toDate == null || !t.getTrainingDate().isAfter(toDate)))
+                .filter(t -> traineeName == null || t.getTrainee().getUser().getUsername().equalsIgnoreCase(traineeName))
+                .map(t -> {
```

Comment:
this logic appears really strange here, selection should be fully made on repository level by username. Additional processing of selection is extra